sirepo/job_driver/__init__.py

```diff
@@ -285,41 +295,36 @@ def _agent_env(self, env=None):
         )
 
     async def _agent_start(self, op):
-        if self._agent_starting:
+        if self._agent_starting_timeout:
             return
         async with self._agent_start_lock:
-            if self._agent_starting or self._websocket_ready.is_set():
+            if self._agent_starting_timeout or self._websocket_ready.is_set():
                 return
             try:
-                self._agent_starting = True
-                # TODO(e-carlin): We need a timeout on agent starts. If an agent
-                # is started but never connects we will be in the '_agent_starting'
-                # state forever. After a timeout we should kill the misbehaving
-                # agent and start a new one.
-                await self.kill()
-                # this starts the process, but _receive_alive sets it to false
-                # when the agent fully starts.
                 pkdlog('{} {} await _do_agent_start', self, op)
+                # All awaits must be after this. If a call hangs the timeout
+                # handler will cancel this task
                 self._agent_starting_timeout = tornado.ioloop.IOLoop.current().call_later(
                     self._AGENT_STARTING_SECS,
                     self._agent_starting_timeout_handler,
                 )
+                # POSIT: CancelledError isn't smothered by any of the below calls
+                await self.kill()
                 await self._do_agent_start(op)
             except Exception as e:
-                pkdlog('{} exception={}', self, e)
-                self._agent_starting_done()
+                pkdlog('{} error={} stack={}', self, e, pkdexc())
+                self.free_resources()
```

Comment:
I had that initially but since the error is logged here and not used once we pass it to free_resources I removed it. If we decide to log the [error in the history](https://github.com/radiasoft/sirepo/pull/2422#discussion_r406805301) then it will be useful and I'll add it back in.

Reply:
I thought we were going to save it in the job file?

Reply:
Not that I recall. We had discussed using it as a flag in the supervisor to differentiate types of cancelledErrors. That became unnecessary so I removed it. Happy to add passing the error along back in.  

If we add it back in:
1. Should we store it in `self.db.error` or somewhere else? 
2. Do we need to differentiate between cancelled due to timeout and cancelled due to api_runCancel?

Reply:
I think we do need to differentiate between canceled due to timeout so we can audit. If a user says "my job got canceled!", and we can look to see if they canceled it or if it was a timeout or if the supervisor was restarted (different case) or if NERSC gave a timeout. We don't need to do all these now, but we should know why a job terminated.